src/tests/ftest/io/nvme_io.py

```
@@ -24,81 +24,52 @@
-        self.context = DaosContext(build_paths['PREFIX'] + '/lib/')
-        self.d_log = DaosLog(self.context)
-        self.hostlist = self.params.get("servers", '/run/hosts/*')
-        self.hostfile = write_host_file.write_host_file(self.hostlist,
-                                                        self.workdir)
-        self.hostlist_clients = self.params.get("clients", '/run/hosts/*')
+        self.hostlist_servers = self.params.get("test_machines", '/run/hosts/*')
         #This is for NVMe Setup
         self.nvme_parameter = self.params.get("bdev_class", '/server_config/')
-
         try:
             if self.nvme_parameter == "nvme":
-                spdk.nvme(self.hostlist, "setup")
-        except:
-            raise
-
-        self.agent_sessions = AgentUtils.run_agent(self.basepath,
-                                                   self.hostlist,
-                                                   self.hostlist_clients)
-
-        #Start Server
-        server_utils.run_server(self.hostfile, self.server_group, self.basepath)
+                spdk.nvme_setup(self.hostlist_servers)
+        except spdk.SpdkFailed as error:
+            self.fail("Error setting up NVMe: {}".format(error))
+        super(NvmeIo, self).setUp()
 
     def tearDown(self):
-        try:
-            if self.pool_connect:
+        if self.pool_connect:
+            try:
                 self.pool.disconnect()
                 self.pool.destroy(1)
+            finally:
+                pass
+
+        try:
+            super(NvmeIo, self).tearDown()
         finally:
-            if self.agent_sessions:
-                AgentUtils.stop_agent(self.hostlist_clients,
-                                      self.agent_sessions)
-            server_utils.stop_server(hosts=self.hostlist)
-            #For NVMe Cleanup
             if self.nvme_parameter == "nvme":
-                spdk.nvme(self.hostlist)
+                spdk.nvme_cleanup(self.hostlist_servers)
```

I think you want this all inside two `try:`/`finally:`s here so that `super(NvmeIo, self).tearDown()` and `spdk.nvme_cleanup(self.hostlist_servers)` both get called no matter what happens before them.

---

I am not sure I followed your comment. It is going to called super(NvmeIo, self).tearDown() and nvme_cleanup() no matter what happens before them.

---

I don't think it does:

```python
try:
    print 10/0
finally:
    pass

print "got here"
```
produces:

```
Traceback (most recent call last):
  File "/tmp/a.py", line 2, in <module>
    print 10/0
ZeroDivisionError: integer division or modulo by zero
```

It didn't get to the print statement.

I suspect you are thinking of something along the lines of:

```python
try:
    print 10/0
finally:
    print "got here"
```

which produces:

```
got here
Traceback (most recent call last):
  File "/tmp/a.py", line 2, in <module>
    print 10/0
ZeroDivisionError: integer division or modulo by zero
```

So you need these two `try:`/`finally:`s to cascade.

---

Ah finally is : pass, I missed that one. I though it's just skip the exception. will update the code

---

Instead of using another try/final block may be better to use except instead of finally for pool.disconnect.

if self.pool_connect:
			    try:
                self.pool.disconnect()
                self.pool.destroy(1)
            except:
                pass

---

But then you are ignoring exceptions in `self.pool.d{isconnect,estroy}()`.

---

yes, if some thing goes wrong in pool d{isconnect,estroy}() we wanted to make sure to stop the servers and clean NVMe no matter what happen. Not sure if we want to catch exception and do some thing else.

---

So in that case, you cannot simply `pass` on an `except:`ion here then.  You either need to catch it and do something, or leave it uncaught so that it gets raised to the caller.  But in either case you want the second `try:`/`finally:` block to happen so it has to be inside the `finally:` of the first block.